### PR TITLE
Implementation of the ranges (with delimiters)

### PR DESCRIPTION
**Pull request by Mingun**
Reincarnation of the https://github.com/pegjs/pegjs/pull/209.
Previous attempt, closed by technical reasons: #208

This is an implementation of the ranges proposal with batteries, i.e.:

- ability to use numeric constants to specify minimum, maximum or exact repetition count:
  ```pegjs
  more2 = "a"|2..|;
  upTo3 = "a"|..3|;
  ```
- ability to use preceding label as a range boundary:
  ```pegjs
  list = count:n5 @n5|count|;
  n5 = n:[0-9]|5| { return parseInt(n); };
  ```
- ability to use function as a boundary:
  ```pegjs
  list = "a"|{ return options.listSize; }|;
  ```

The syntax chosen saves the `<` and `>` characters for the template definitions where their are more natural.
`[` and `]` already used for character class definitions and 
`(` and `)` already used for grouping.

### CODE REVIEW

**Comment by polkovnikov-ph** [comment]
What if there was a way to define parser-functions, and it would be just `repeat(x, 2..)` and `repeat(x, ..3)`?

I think by creating a new syntax for every feature it will resemble Perl and its regular expressions quite soon (arguably, it already resembles it way too much).

The thing not mentioned in PR description, but probably the most awaited by me, is delimited repetition (`a|.., b|`).
1. Is there a CI version of this PR to play with it?
2. What AST does it generate?
3. I remember there was a discussion in PEG.js issues to add it as `a % b` operator. Does this PR subsume that feature request?
4. I didn't find tests attempting to break parsing of `||` brackets. Can we be sure that some `.. | .. | | .. , .. | .. , .. | |` will get parsed properly?

**Comment by Mingun** [comment]
1. You can checkout this branch, build minified version of a parser
    ```console
    npm run build
    # or only
    npm run rollup
    npm run terser
    npm run deploy
    ```
    and go to `docs/online.html`:
3. You can use a recently added `peggy --ast` option to look at it
4. I personally against any syntax which does not clearly indicate where the repeat expression ends. Using `a % b` falls into this category. Last time I tried to summarize possible options [here](https://github.com/peggyjs/peggy/pull/208#issuecomment-1140322295)
5. Yes, but you should take into account, that you cannot put two suffix operators one after another, you must wrap the first in parentheses:
    ```pegjs
    // This is all forbidden
    start1 = .. | .. | | .. , .. | .. , .. | |;
    start2 = .**;
    start3 = .++;
    start4 = .??;

    // This will work
    start5 = .(. | .. |) | .. , .. | .. , .. | |;
    start6 = (.*)*;
    start7 = (.+)+;
    start8 = (.?)?;
    ```

**Comment by hildjj** [comment]
> What if there was a way to define parser-functions

I think we will soon need to determine what our long-term syntax extensibility approach is, before we run out of adequate syntax to act as the extensibility point.  I don't think `repeat(n)` will work, since that looks like a rule named `repeat` followed by a group.  `<repeat min="a" max="b">...</repeat>` would work, but that's pretty ugly, and even I don't like XML that much.

File: lib/peg.d.ts [hunk 1]
```diff
@@ -178,19 +182,63 @@ declare namespace ast {
...
+  interface Repeated extends Expr<"repeated"> {
+    /**
+     * Minimum count of repetitions. If `null` then exact repetition
+     * is used and minimum the same as maximum.
+     */
+    min: RepeatedBoundary | null;
+    /** Maximum count of repetitions. */
+    max: RepeatedBoundary;
+    /**
+     * An expression that should appeared between occurrences of the `expression`.
```
Comment (hildjj) [inline lib/peg.d.ts:235]: "appeared" -> "appear"

File: lib/peg.d.ts [hunk 2]
```diff
@@ -607,6 +655,13 @@ export namespace compiler {
        * @param args Any arguments passed to the `Visitor`
        */
       one_or_more?(node: ast.Suffixed, ...args: any[]): any;
+      /**
+       * Default behavior: run visitor on `expression` and return it result
```
Comment (hildjj) [inline lib/peg.d.ts:659]: It looks like the default implementation consumes the delimiter if it exists, then consumes the expression.

File: docs/js/examples.peggy [hunk 3]
```diff
@@ -16,6 +16,8 @@ star = match:"a"* rest:rest { return {match, rest}; }
 
 plus = match:"a"+ rest:rest { return {match, rest}; }
 
+repetition = match:"a"|2..3, ","| rest:rest { return {match, rest}; }
```
Comment (hildjj) [inline docs/js/examples.peggy:19]: Playing around with this, it looks like this doesn't work as expected:

```peg.js
foo = n:num @"a" |n..n, "b" |n..n||

num = n:[0-9]+ { return parseInt(n, 10); }
```

Inputs: "2aba" matches, and I expected it to fail wanting another "b" before the last "a".  I assume this means that the inner range doesn't have visibility to the names from the outer block.  This isn't a big problem, but we should either fix it, find a work-around, or document it.

File: src/parser.pegjs [hunk 4]
```diff
@@ -430,6 +470,9 @@ BareCodeBlock
 Code
   = $((![{}] SourceCharacter)+ / "{" Code "}")*
 
+Integer
+  = digits:$DecimalDigit+ { return parseInt(digits); }
```
Comment (hildjj) [inline src/parser.pegjs:474]: I like to make the `10` explicit in parseInt, but don't bother to fix unless you agree.

**Review by hildjj — COMMENTED** [review]
I'm sorry this took so long to get to.

The only substantive thing here is to figure out what happens with names being passed to nested repeats.

Once we get that nailed, and get this rebased, I'm ready to land it and start working toward a release.

**Comment by hildjj** [comment]
@Mingun do you have time to work on this?  If not, I'll try to rebase it at least.

**Comment by Mingun** [comment]
I'll look at weekend

File: docs/documentation.html [hunk 5]
```diff
@@ -832,6 +832,63 @@ <h3 id="grammar-syntax-and-semantics-parsing-expression-types">Parsing Expressio
...
+             <code>"a,a,a,"</code>, <code>"a,a,a,a"</code></div>
+      </div>
+      <div class="try">
+        <em>Try it:</em>
+        <input type="text" value="a,a" class="exampleInput" name="repetition">
+        <div class="result"></div>
+      </div>
+    </div>
+  </dd>
+
```
[the remark this answers is not in the document]
Comment (Mingun) [reply docs/documentation.html:891]: Updated

File: docs/js/examples.peggy (hunk 3 above)
Comment (Mingun) [reply docs/js/examples.peggy:19]: This is a bug, looking for it...

Comment (Mingun) [reply docs/js/examples.peggy:19]: Seems to be fixed, but I want to check if more tests in parser behavior section is needed

**Comment by Mingun** [comment]
Ok, the bug seems to be fixed, I'll check it once again on this week and add a changelog entry for plugins' authors.

**Comment by Mingun** [comment]
The diff between the 7-month old version and the current:
- added forgotten tests for function-based boundaries
- fixes some minor errors in tests for variable-based boundaries
- added a notice for plugins' authors
- fix some minor misprints in tests

The last thing remains is to add tests for source-mapping functions from function-based boundaries

File: lib/compiler/opcodes.js [hunk 6]
```diff
@@ -26,6 +26,10 @@ const opcodes = {
   IF:                13,   // IF t, f
   IF_ERROR:          14,   // IF_ERROR t, f
   IF_NOT_ERROR:      15,   // IF_NOT_ERROR t, f
+  IF_LT:             30,   // IF_LT min, t, f
```
[the remark this answers is not in the document]
Comment (Mingun) [reply lib/compiler/opcodes.js:29]: Added

File: docs/js/examples.peggy (hunk 3 above)
Comment (Mingun) [reply docs/js/examples.peggy:19]: Added checks to `generate-bytecode.spec.js` and `generated-parser-behavior.spec.js`.

**Comment by hildjj** [comment]
This is *fantastic* work.  Thank you very much.

I'm going to merge  as-is, then when I do a full review of the documentation page, there may be a few nits I clean up while fixing other things.
